File: cpp/leetcode/059.cpp
class Solution {
public:
    vector<vector<int>> generateMatrix(int n) {
        int cnts = 1;
        int m = n, t = 0, l = 0;
        vector<vector<int>> matrix(n, vector<int>(n, 0));
        while (t < m && l < n) {
            for (int i = l; i < n; i++) {
                matrix[t][i] = cnts++;
            }
            t++;
            for (int i = t; i < m; i++) {
                matrix[i][n - 1] = cnts++;
            }
            n--;
            if (t < m) {
                for (int i = n - 1; i >= l; i--) {
                    matrix[m - 1][i] = cnts++;
                }
                m--;
            }
            if (l < n) {
                for (int i = m - 1; i >= t; i--) {
                    matrix[i][l] = cnts++;
                }
                l++;
            }
        }
        return matrix;
    }
};

File: python/leetcode/059.py
class Solution(object):
    def generateMatrix(self, n):
        res = [[0] * n for i in range(n)]
        k, l, m = 0, 0, n
        cnts = 1
        while k < m and l < n:
            for i in range(l, n):
                res[k][i] = cnts
                cnts += 1
            k += 1
            for i in range(k, m):
                res[i][n - 1] = cnts  # n - 1
                cnts += 1
            n -= 1
            if k < m:  # k < m
                for i in range(n - 1, l - 1, -1):  # n - 1
                    res[m - 1][i] = cnts  # m - 1
                    cnts += 1
                m -= 1
            if l < n:  # l < n
                for i in range(m - 1, k - 1, -1):  # m - 1
                    res[i][l] = cnts
                    cnts += 1
                l += 1
        return res